src/WebJobs.Script.WebHost/Diagnostics/ConsoleWriter.cs

```diff
@@ -0,0 +1,188 @@
+            if (_consoleBuffer.Writer.TryWrite(evt) == false)
+            {
+                Console.WriteLine(evt);
+            }
+        }
+
+        internal void StartProcessingBuffer(bool batched)
+        {
+            // intentional no-op if the task is already running
+            if (_consoleBufferReadLoop == null || _consoleBufferReadLoop.IsCompleted)
+            {
+                _consoleBufferReadLoop = ProcessConsoleBufferAsync(batched);
+            }
+        }
+
+        private async Task ProcessConsoleBufferAsync(bool batched)
+        {
+            try
+            {
+                if (batched)
+                {
+                    await ProcessConsoleBufferBatchedAsync();
+                }
+                else
+                {
+                    await ProcessConsoleBufferNonBatchedAsync();
+                }
+            }
+            catch (Exception ex)
+            {
+                // Not sure what can fail here, but there is nothing else monitoring this task so just log the exception
+                _exceptionhandler(ex);
+            }
+            finally
+            {
+                // if this has failed for any reason, fall everything back to console
+                _writeEvent = Console.WriteLine;
+                _consoleBuffer.Writer.TryComplete();
+            }
+        }
+
+        private async Task ProcessConsoleBufferNonBatchedAsync()
+        {
+            await foreach (var line in _consoleBuffer.Reader.ReadAllAsync())
+            {
+                Console.WriteLine(line);
+            }
+        }
+
+        private async Task ProcessConsoleBufferBatchedAsync()
+        {
+            var builder = new StringBuilder();
+
+            while (await _consoleBuffer.Reader.WaitToReadAsync())
+            {
+                if (_consoleBuffer.Reader.TryRead(out string line1))
+                {
+                    // Can we synchronously read multiple lines?
+                    // If yes, use the string builder to batch them together into a single write
+                    // If no, just write the single line without using the builder;
+                    if (_consoleBuffer.Reader.TryRead(out string line2))
+                    {
+                        builder.AppendLine(line1);
+                        builder.AppendLine(line2);
+                        int lines = 2;
+
+                        while (lines < DefaultBufferSize && _consoleBuffer.Reader.TryRead(out string nextLine))
+                        {
+                            builder.AppendLine(nextLine);
+                            lines++;
+                        }
+
+                        Console.Write(builder.ToString());
+                        builder.Clear();
+                    }
+                    else
+                    {
+                        Console.WriteLine(line1);
+                    }
+                }
+            }
+        }
+
+        protected virtual void Dispose(bool disposing)
```

Comment:
You can implement `IAsyncDisposable` instead if you want access to async.

Reply:
Good suggestion! The only implementor in the codebase currently is AspNetCoreGrpcServer, and it happens to implement both IDisposable and IAsyncDisposable. Do you know if this is required? The sync implementation is quite odd in that it calls the async implementation without blocking on the valuetask, so as far as I can tell it breaks the expected IDisposable contract (disposing might not be done by the time the method returns).
https://github.com/Azure/azure-functions-host/blob/4ec73a34dea76db91b6861db59d57d1ff00265f8/src/WebJobs.Script.Grpc/Server/AspNetCoreGrpcServer.cs#L60-L82

Reply:
@jviau Unfortunately, I can't get this to work. I tested by using ctrl+c to shutdown a running host, and the Dispose implementation runs but the DisposeAync implementation does not. Either there is something wrong with my testing methodology, or our DI container doesn't have support for IAsyncDisposable.

Reply:
Sticking with IDispose implementation for now.